File: server/src/routes/solutions.js
const express = require('express');
const prisma = require('../db');
const { authenticateUser } = require('./auth');
const router = express.Router();

// POST /api/solutions/sync (Extension background worker auto-push)
router.post('/sync', authenticateUser, async (req, res) => {
  const { solutions } = req.body;
  if (!solutions || !Array.isArray(solutions)) {
    return res.status(400).json({ error: 'Payload must contain a "solutions" array.' });
  }

  const defaultUserId = req.user.id;
  const syncResults = [];

  try {
    for (const sol of solutions) {
      const submissionId = String(sol.submissionId || sol.id || `${sol.challengeSlug}_${sol.language}_${sol.submittedAt || Date.now()}`);
      const challengeSlug = sol.challengeSlug || sol.slug || 'unknown';
      const challengeTitle = sol.challengeTitle || sol.name || challengeSlug;
      const contestSlug = sol.contestSlug || 'master';
      const language = sol.language || 'javascript';
      const code = sol.code || '';
      const score = typeof sol.score === 'number' ? sol.score : parseFloat(sol.score) || 1.0;
      const status = sol.status || 'Accepted';
      const submittedAt = sol.submittedAt ? new Date(sol.submittedAt) : new Date();

      if (!code) continue;

      // Determine target user for this solution
      let targetUserId = defaultUserId;
      if (sol.username && typeof sol.username === 'string') {
        const usernameClean = sol.username.trim();
        let targetUser = await prisma.user.findUnique({ where: { username: usernameClean } });
        if (!targetUser) {
          targetUser = await prisma.user.create({
            data: {
              username: usernameClean,
              token: `hr_${usernameClean}_${Math.random().toString(36).substring(2, 8)}`,
              role: 'USER'
            }
          });
        }
        targetUserId = targetUser.id;
      }

      // If this submissionId already exists under another user (e.g. admin), remove old duplicate
      const existingAdminSol = await prisma.solution.findFirst({
        where: {
          submissionId,
          user: { role: 'ADMIN' }
        }
      });
      if (existingAdminSol && existingAdminSol.userId !== targetUserId) {
        await prisma.solution.delete({ where: { id: existingAdminSol.id } }).catch(() => {});
      }

      const record = await prisma.solution.upsert({
        where: {
          submissionId_userId: { submissionId, userId: targetUserId }
        },
        update: {
          challengeTitle,
          contestSlug,
          language,
          code,
          score,
          status,
          submittedAt
        },
        create: {
          submissionId,
          challengeSlug,
          challengeTitle,
          contestSlug,
          language,
          code,
          score,
          status,
          submittedAt,
          userId: targetUserId
        }
      });
      syncResults.push(record.id);
    }

    // Auto-remap legacy admin-owned solutions where submissionId encodes username (e.g. slug_username)
    const adminUser = await prisma.user.findUnique({ where: { token: 'hr_admin_master_token_2026' } });
    if (adminUser) {
      const adminSolutions = await prisma.solution.findMany({ where: { userId: adminUser.id } });
      for (const sol of adminSolutions) {
        const lastUnderscoreIdx = sol.submissionId.lastIndexOf('_');
        if (lastUnderscoreIdx > 0) {
          const possibleUsername = sol.submissionId.substring(lastUnderscoreIdx + 1);
          if (possibleUsername && possibleUsername !== 'admin') {
            let userMatch = await prisma.user.findUnique({ where: { username: possibleUsername } });
            if (!userMatch) {
              userMatch = await prisma.user.create({
                data: {
                  username: possibleUsername,
                  token: `hr_${possibleUsername}_${Math.random().toString(36).substring(2, 8)}`,
                  role: 'USER'
                }
              });
            }
            await prisma.solution.update({
              where: { id: sol.id },
              data: { userId: userMatch.id }
            }).catch(() => {});
          }
        }
      }
    }

    res.json({
      success: true,
      syncedCount: syncResults.length,
      syncedUser: req.user.username
    });
  } catch (err) {
    console.error('Error syncing solutions:', err);
    res.status(500).json({ error: err.message });
  }
});

// GET /api/solutions/users (List users for filters)
router.get('/users', async (req, res) => {
  try {
    const users = await prisma.user.findMany({
      select: {
        id: true,
        username: true,
        discordAvatar: true,
        role: true,
        _count: { select: { solutions: true } }
      },
      orderBy: { username: 'asc' }
    });
    res.json({ users });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// GET /api/solutions (Query list)
router.get('/', async (req, res) => {
  const { challengeSlug, language, userId, userIds, username, usernames, search, page = 1, limit = 50 } = req.query;

  const where = {};
  if (challengeSlug) where.challengeSlug = challengeSlug;
  if (language) where.language = language;

  const rawUsernames = usernames || username;
  const rawUserIds = userIds || userId;

  if (rawUsernames) {
    const nameList = (Array.isArray(rawUsernames) ? rawUsernames : String(rawUsernames).split(',')).map(s => s.trim()).filter(Boolean);
    if (nameList.length === 1) {
      where.user = { username: nameList[0] };
    } else if (nameList.length > 1) {
      where.user = { username: { in: nameList } };
    }
  } else if (rawUserIds) {
    const idList = (Array.isArray(rawUserIds) ? rawUserIds : String(rawUserIds).split(',')).map(s => s.trim()).filter(Boolean);
    if (idList.length === 1) {
      where.userId = idList[0];
    } else if (idList.length > 1) {
      where.userId = { in: idList };
    }
  }

  if (search) {
    where.OR = [
      { challengeTitle: { contains: search } },
      { challengeSlug: { contains: search } },
      { code: { contains: search } }
    ];
  }

  const skip = (parseInt(page) - 1) * parseInt(limit);
  const take = parseInt(limit);

  try {
    const [solutions, totalCount] = await Promise.all([
      prisma.solution.findMany({
        where,
        skip,
        take,
        orderBy: { updatedAt: 'desc' },
        include: {
          user: { select: { id: true, username: true, discordAvatar: true, role: true } },
          ratings: true,
          _count: { select: { comments: true, reviewRounds: true } }
        }
      }),
      prisma.solution.count({ where })
    ]);

    // Attach calculated average ratings
    const formattedSolutions = solutions.map(sol => {
      const clevernessAvg = sol.ratings.length > 0
        ? (sol.ratings.reduce((acc, r) => acc + r.cleverness, 0) / sol.ratings.length).toFixed(1)
        : null;
      const readabilityAvg = sol.ratings.length > 0
        ? (sol.ratings.reduce((acc, r) => acc + r.readability, 0) / sol.ratings.length).toFixed(1)
        : null;

      return {
        ...sol,
        clevernessAvg: clevernessAvg ? parseFloat(clevernessAvg) : null,
        readabilityAvg: readabilityAvg ? parseFloat(readabilityAvg) : null,
        ratingsCount: sol.ratings.length,
        ratings: undefined // omit raw array in list view for performance
      };
    });

    res.json({
      solutions: formattedSolutions,
      pagination: {
        total: totalCount,
        page: parseInt(page),
        limit: parseInt(limit),
        totalPages: Math.ceil(totalCount / limit)
      }
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// GET /api/solutions/:id (Single solution details)
router.get('/:id', async (req, res) => {
  try {
    const solution = await prisma.solution.findUnique({
      where: { id: req.params.id },
      include: {
        user: { select: { id: true, username: true, discordUsername: true, discordAvatar: true, role: true } },
        ratings: {
          include: { user: { select: { id: true, username: true } } }
        },
        comments: {
          orderBy: { createdAt: 'asc' },
          include: { user: { select: { id: true, username: true, discordAvatar: true, role: true } } }
        },
        reviewRounds: {
          orderBy: { roundNumber: 'asc' },
          include: { reviewer: { select: { id: true, username: true, role: true } } }
        }
      }
    });

    if (!solution) {
      return res.status(404).json({ error: 'Solution not found' });
    }

    const clevernessAvg = solution.ratings.length > 0
      ? (solution.ratings.reduce((acc, r) => acc + r.cleverness, 0) / solution.ratings.length).toFixed(1)
      : null;
    const readabilityAvg = solution.ratings.length > 0
      ? (solution.ratings.reduce((acc, r) => acc + r.readability, 0) / solution.ratings.length).toFixed(1)
      : null;

    res.json({
      solution: {
        ...solution,
        clevernessAvg: clevernessAvg ? parseFloat(clevernessAvg) : null,
        readabilityAvg: readabilityAvg ? parseFloat(readabilityAvg) : null
      }
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: server/handlers/solutions.go
package handlers

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"gorm.io/gorm"
	"hackerrank-server/db"
	"hackerrank-server/models"
)

type SyncSolutionItem struct {
	ID             string      `json:"id"`
	SubmissionID   string      `json:"submissionId"`
	ChallengeSlug  string      `json:"challengeSlug"`
	ChallengeTitle string      `json:"challengeTitle"`
	Slug           string      `json:"slug"`
	Name           string      `json:"name"`
	ContestSlug    string      `json:"contestSlug"`
	Language       string      `json:"language"`
	Code           string      `json:"code"`
	Score          interface{} `json:"score"`
	Status         string      `json:"status"`
	SubmittedAt    interface{} `json:"submittedAt"`
	Username       string      `json:"username"`
}

func SolutionsSyncHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	currentUser, _ := GetUserFromContext(r.Context())

	var payload struct {
		Solutions []SyncSolutionItem `json:"solutions"`
	}

	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil || payload.Solutions == nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"error": "Payload must contain a \"solutions\" array."})
		return
	}

	syncedIDs := make([]string, 0)

	for _, sol := range payload.Solutions {
		code := strings.TrimSpace(sol.Code)
		if code == "" {
			continue
		}

		challengeSlug := sol.ChallengeSlug
		if challengeSlug == "" {
			challengeSlug = sol.Slug
		}
		if challengeSlug == "" {
			challengeSlug = "unknown"
		}

		challengeTitle := sol.ChallengeTitle
		if challengeTitle == "" {
			challengeTitle = sol.Name
		}
		if challengeTitle == "" {
			challengeTitle = challengeSlug
		}

		submissionID := sol.SubmissionID
		if submissionID == "" {
			submissionID = sol.ID
		}
		if submissionID == "" {
			submissionID = fmt.Sprintf("%s_%s_%d", challengeSlug, sol.Language, time.Now().UnixNano())
		}

		contestSlug := sol.ContestSlug
		if contestSlug == "" {
			contestSlug = "master"
		}

		language := sol.Language
		if language == "" {
			language = "javascript"
		}

		status := sol.Status
		if status == "" {
			status = "Accepted"
		}

		var scoreVal float64 = 1.0
		if sol.Score != nil {
			switch v := sol.Score.(type) {
			case float64:
				scoreVal = v
			case string:
				if parsed, err := strconv.ParseFloat(v, 64); err == nil {
					scoreVal = parsed
				}
			}
		}

		submittedAt := time.Now()
		if sol.SubmittedAt != nil {
			if str, ok := sol.SubmittedAt.(string); ok && str != "" {
				if parsedTime, err := time.Parse(time.RFC3339, str); err == nil {
					submittedAt = parsedTime
				}
			}
		}

		targetUserID := currentUser.ID
		if strings.TrimSpace(sol.Username) != "" {
			cleanUsername := strings.TrimSpace(sol.Username)
			var targetUser models.User
			if err := db.DB.Where("username = ?", cleanUsername).First(&targetUser).Error; err != nil {
				randBytes := make([]byte, 4)
				rand.Read(randBytes)
				newToken := fmt.Sprintf("hr_%s_%s", cleanUsername, hex.EncodeToString(randBytes))
				targetUser = models.User{
					ID:       uuid.New().String(),
					Username: cleanUsername,
					Token:    newToken,
					Role:     "USER",
				}
				db.DB.Create(&targetUser)
			}
			targetUserID = targetUser.ID
		}

		// Clean up old admin duplicate if target user is different
		var existingAdminSol models.Solution
		if err := db.DB.Joins("User").Where("Solution.submissionId = ? AND User.role = 'ADMIN'", submissionID).First(&existingAdminSol).Error; err == nil {
			if existingAdminSol.UserID != targetUserID {
				db.DB.Where("id = ?", existingAdminSol.ID).Delete(&models.Solution{})
			}
		}

		var record models.Solution
		err := db.DB.Where("submissionId = ? AND userId = ?", submissionID, targetUserID).First(&record).Error
		if err != nil {
			// Create new solution
			record = models.Solution{
				ID:             uuid.New().String(),
				SubmissionID:   submissionID,
				ChallengeSlug:  challengeSlug,
				ChallengeTitle: challengeTitle,
				ContestSlug:    &contestSlug,
				Language:       language,
				Code:           code,
				Score:          &scoreVal,
				Status:         &status,
				SubmittedAt:    &submittedAt,
				UserID:         targetUserID,
			}
			db.DB.Create(&record)
		} else {
			// Update existing solution
			db.DB.Model(&record).Updates(map[string]interface{}{
				"challengeTitle": challengeTitle,
				"contestSlug":    contestSlug,
				"language":       language,
				"code":           code,
				"score":          scoreVal,
				"status":         status,
				"submittedAt":    submittedAt,
			})
		}

		syncedIDs = append(syncedIDs, record.ID)
	}

	// Auto-remap legacy admin-owned solutions where submissionId encodes username
	var adminUser models.User
	if err := db.DB.Where("token = ?", "hr_admin_master_token_2026").First(&adminUser).Error; err == nil {
		var adminSols []models.Solution
		db.DB.Where("userId = ?", adminUser.ID).Find(&adminSols)
		for _, sol := range adminSols {
			idx := strings.LastIndex(sol.SubmissionID, "_")
			if idx > 0 {
				possibleUsername := sol.SubmissionID[idx+1:]
				if possibleUsername != "" && possibleUsername != "admin" {
					var userMatch models.User
					if err := db.DB.Where("username = ?", possibleUsername).First(&userMatch).Error; err != nil {
						randBytes := make([]byte, 4)
						rand.Read(randBytes)
						newToken := fmt.Sprintf("hr_%s_%s", possibleUsername, hex.EncodeToString(randBytes))
						userMatch = models.User{
							ID:       uuid.New().String(),
							Username: possibleUsername,
							Token:    newToken,
							Role:     "USER",
						}
						db.DB.Create(&userMatch)
					}
					db.DB.Model(&sol).Update("userId", userMatch.ID)
				}
			}
		}
	}

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success":     true,
		"syncedCount": len(syncedIDs),
		"syncedUser":  currentUser.Username,
	})
}

func SolutionsGetUsersHandler(w http.ResponseWriter, r *http.Request) {
	GetUsersListPublicHandler(w, r)
}

func SolutionsListHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	q := r.URL.Query()

	challengeSlug := q.Get("challengeSlug")
	language := q.Get("language")
	search := q.Get("search")

	rawUsernames := q.Get("usernames")
	if rawUsernames == "" {
		rawUsernames = q.Get("username")
	}

	rawUserIDs := q.Get("userIds")
	if rawUserIDs == "" {
		rawUserIDs = q.Get("userId")
	}

	limitStr := q.Get("limit")
	limit := 50
	if limitStr != "" {
		if l, err := strconv.Atoi(limitStr); err == nil && l > 0 {
			limit = l
		}
	}

	pageStr := q.Get("page")
	page := 1
	if pageStr != "" {
		if p, err := strconv.Atoi(pageStr); err == nil && p > 0 {
			page = p
		}
	}

	offsetStr := q.Get("offset")
	skip := (page - 1) * limit
	if offsetStr != "" {
		if off, err := strconv.Atoi(offsetStr); err == nil && off >= 0 {
			skip = off
		}
	}

	query := db.DB.Model(&models.Solution{}).Joins("User")

	if challengeSlug != "" {
		query = query.Where("Solution.challengeSlug = ?", challengeSlug)
	}
	if language != "" {
		query = query.Where("Solution.language = ?", language)
	}

	if rawUsernames != "" {
		parts := strings.Split(rawUsernames, ",")
		names := make([]string, 0, len(parts))
		for _, p := range parts {
			if trimmed := strings.TrimSpace(p); trimmed != "" {
				names = append(names, trimmed)
			}
		}
		if len(names) == 1 {
			query = query.Where("User.username = ?", names[0])
		} else if len(names) > 1 {
			query = query.Where("User.username IN ?", names)
		}
	} else if rawUserIDs != "" {
		parts := strings.Split(rawUserIDs, ",")
		ids := make([]string, 0, len(parts))
		for _, p := range parts {
			if trimmed := strings.TrimSpace(p); trimmed != "" {
				ids = append(ids, trimmed)
			}
		}
		if len(ids) == 1 {
			query = query.Where("Solution.userId = ?", ids[0])
		} else if len(ids) > 1 {
			query = query.Where("Solution.userId IN ?", ids)
		}
	}

	if search != "" {
		searchPattern := "%" + search + "%"
		query = query.Where("Solution.challengeTitle LIKE ? OR Solution.challengeSlug LIKE ? OR Solution.code LIKE ?", searchPattern, searchPattern, searchPattern)
	}

	var totalCount int64
	query.Count(&totalCount)

	var solutions []models.Solution
	query.Order("Solution.updatedAt desc").Limit(limit).Offset(skip).Preload("User").Preload("Ratings").Find(&solutions)

	type FormattedSolution struct {
		models.Solution
		ClevernessAvg  *float64 `json:"clevernessAvg"`
		ReadabilityAvg *float64 `json:"readabilityAvg"`
		RatingsCount   int      `json:"ratingsCount"`
		Count          struct {
			Comments     int64 `json:"comments"`
			ReviewRounds int64 `json:"reviewRounds"`
		} `json:"_count"`
	}

	formatted := make([]FormattedSolution, 0, len(solutions))
	for _, sol := range solutions {
		var cAvg, rAvg *float64
		if len(sol.Ratings) > 0 {
			var cSum, rSum int
			for _, r := range sol.Ratings {
				cSum += r.Cleverness
				rSum += r.Readability
			}
			cVal := math.Round((float64(cSum)/float64(len(sol.Ratings)))*10) / 10
			rVal := math.Round((float64(rSum)/float64(len(sol.Ratings)))*10) / 10
			cAvg = &cVal
			rAvg = &rVal
		}

		var commentCount, reviewCount int64
		db.DB.Model(&models.Comment{}).Where("solutionId = ?", sol.ID).Count(&commentCount)
		db.DB.Model(&models.ReviewRound{}).Where("solutionId = ?", sol.ID).Count(&reviewCount)

		item := FormattedSolution{
			Solution:       sol,
			ClevernessAvg:  cAvg,
			ReadabilityAvg: rAvg,
			RatingsCount:   len(sol.Ratings),
		}
		item.Ratings = nil // Omit raw ratings array in list view for performance
		item.Count.Comments = commentCount
		item.Count.ReviewRounds = reviewCount

		formatted = append(formatted, item)
	}

	hasMore := (skip + len(formatted)) < int(totalCount)
	nextOffset := skip + len(formatted)

	totalPages := int(math.Ceil(float64(totalCount) / float64(limit)))
	if totalPages < 0 {
		totalPages = 0
	}

	json.NewEncoder(w).Encode(map[string]interface{}{
		"solutions": formatted,
		"pagination": map[string]interface{}{
			"total":      totalCount,
			"offset":     skip,
			"limit":      limit,
			"page":       page,
			"totalPages": totalPages,
			"hasMore":    hasMore,
			"nextOffset": nextOffset,
		},
	})
}

func SolutionsGetSingleHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	id := chi.URLParam(r, "id")

	var sol models.Solution
	err := db.DB.Preload("User").
		Preload("Ratings", func(db *gorm.DB) *gorm.DB {
			return db.Preload("User", func(db *gorm.DB) *gorm.DB {
				return db.Select("id, username")
			})
		}).
		Preload("Comments", func(db *gorm.DB) *gorm.DB {
			return db.Order("createdAt asc").Preload("User", func(db *gorm.DB) *gorm.DB {
				return db.Select("id, username, discordAvatar, role")
			})
		}).
		Preload("ReviewRounds", func(db *gorm.DB) *gorm.DB {
			return db.Order("roundNumber asc").Preload("Reviewer", func(db *gorm.DB) *gorm.DB {
				return db.Select("id, username, role")
			})
		}).
		Where("id = ?", id).First(&sol).Error

	if err != nil {
		w.WriteHeader(http.StatusNotFound)
		json.NewEncoder(w).Encode(map[string]string{"error": "Solution not found"})
		return
	}

	var cAvg, rAvg *float64
	if len(sol.Ratings) > 0 {
		var cSum, rSum int
		for _, r := range sol.Ratings {
			cSum += r.Cleverness
			rSum += r.Readability
		}
		cVal := math.Round((float64(cSum)/float64(len(sol.Ratings)))*10) / 10
		rVal := math.Round((float64(rSum)/float64(len(sol.Ratings)))*10) / 10
		cAvg = &cVal
		rAvg = &rVal
	}

	type SingleSolutionWrapper struct {
		models.Solution
		ClevernessAvg  *float64 `json:"clevernessAvg"`
		ReadabilityAvg *float64 `json:"readabilityAvg"`
	}

	json.NewEncoder(w).Encode(map[string]interface{}{
		"solution": SingleSolutionWrapper{
			Solution:       sol,
			ClevernessAvg:  cAvg,
			ReadabilityAvg: rAvg,
		},
	})
}
